EventFilter/Phase2PixelRawToDigi/plugins/Phase2ITQCoreProducer.cc

```diff
@@ -0,0 +1,226 @@
+  typedef math::XYZPointD Point;
+  typedef std::vector<Point> PointCollection;
+};
+
+Phase2ITQCoreProducer::Phase2ITQCoreProducer(const edm::ParameterSet& iConfig)
+    : src_(iConfig.getParameter<edm::InputTag>("src")),
+      pixelDigi_token_(consumes<edm::DetSetVector<PixelDigi>>(iConfig.getParameter<edm::InputTag>("siPixelDigi"))),
+      tTopoToken_(esConsumes<TrackerTopology, TrackerTopologyRcd>()) {
+  produces<edm::DetSetVector<Phase2ITQCore>>();
+  produces<edm::DetSetVector<Phase2ITChipBitStream>>();
+}
+
+namespace {
+  // Dimension for 2 chips module = 672 X 434 = 672 X (216 + 1 + 216 + 1)
+  // Dimension for 4 chips module = 1354 X 434 = (672 + 5 + 672 + 5) X (216 + 1 + 216 + 1)
+  // Spacing 1 in column and 5 in row is introduced for each chip in between
+  // if neighboring chip exists
+  const int kQCoresInChipRow = (672);
+  const int kQCoresInChipColumn = (216);
+  const int kQCoresInChipRowGap = (5);
+  const int kQCoresInChipColumnGap = (10);
+}  // namespace
+
+Phase2ITDigiHit updateHitCoordinatesForLargePixels(Phase2ITDigiHit& hit) {
+  /*
+        In-place modification of Hit coordinates to take into account large pixels
+        Hits corresponding to large pixels are remapped so they lie on the boundary of the chip
+        Note that this operation can produce multiple hits with the same row/column coordinates
+        Duplicates get removed later on
+    */
+
+  // Current values before remapping
+  int row = hit.row();
+  int col = hit.col();
+
+  // Values after remapping
+  int updated_row = 0;
+  int updated_col = 0;
+
+  // Remapping of the row coordinate
+  if (row < kQCoresInChipRow) {
+    updated_row = row;
+  } else if (row < (kQCoresInChipRow + kQCoresInChipRowGap)) {
+    updated_row = kQCoresInChipRow - 1;
+  }  // This will be ignored for 2 chips module
+  else if (row < (kQCoresInChipRow + 2 * kQCoresInChipRowGap)) {
+    updated_row = kQCoresInChipRow;
+  } else {
+    updated_row = (hit.row() - 2 * kQCoresInChipRowGap);
+  }
+
+  // Remapping of the column coordinate
+  if (col < kQCoresInChipColumn) {
+    updated_col = col;
+  } else if (col < kQCoresInChipColumn + kQCoresInChipColumnGap) {
+    updated_col = kQCoresInChipColumn - kQCoresInChipColumnGap;
+  } else if (col < (kQCoresInChipColumn + 2 * kQCoresInChipColumn)) {
+    updated_col = kQCoresInChipColumn;
+  } else {
+    updated_col = (hit.col() - 2 * kQCoresInChipColumnGap);
+  }
+
+  hit.set_row(updated_row);
+  hit.set_col(updated_col);
+
+  return hit;
+}
+
+std::vector<Phase2ITDigiHit> adjustEdges(std::vector<Phase2ITDigiHit> hitList) {
+  /*
+        In-place modification of Hit coordinates to take into account large pixels
+    */
+  std::for_each(hitList.begin(), hitList.end(), &updateHitCoordinatesForLargePixels);
+  return hitList;
+}
+
+std::vector<Phase2ITChip> splitByChip(std::vector<Phase2ITDigiHit> hitList) {
+  // Split the hit list by read out chip
+  std::vector<std::vector<Phase2ITDigiHit>> hits_per_chip(4);
```

Comment:
```suggestion
  std::array<std::vector<Phase2ITDigiHit>, 4> hits_per_chip;
```
would avoid one heap memory allocation per call.